Incremental build tests and improvements

cdmihai opened the pull request:
- add the pdb and the xml to the incremental output
- add the lock file to the incremental inputs (and expose the lock file in the project model)
- make incremental resolution print to stdout
- added tests for more outputs and project dependency graphs

piotrpMSFT commented:
@dotnet-bot test CentOS7.1 Release Build please.

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -123,17 +123,17 @@ private bool NeedsRebuilding(ProjectContext project, ProjectDependenciesFacade d
+            Reporter.Output.WriteLine($"Oldest output item was written at {minDate} : {minOutputPath}");
+            Reporter.Output.WriteLine($"Inputs newer than the oldest output item:");
 
             foreach (var newInput in newInputs)
```
This should probably stay verbose. It's a little noisy

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -148,14 +148,14 @@ private static bool AnyMissingIO(ProjectContext project, IEnumerable<string> ite
-            Reporter.Verbose.WriteLine($"\nProject {project.ProjectName()} will be compiled because expected {itemsType} are missing: ");
+            Reporter.Output.WriteLine($"\nProject {project.ProjectName()} will be compiled because expected {itemsType} are missing: ");
 
             foreach (var missing in missingItems)
```
This as well.

cdmihai replied:
Should I keep any of it non-verbose?

davidfowl replied:
You need to tell the user just enough that something happened. To see why they'll have to use verbose.

cdmihai replied:
In non-verbose output, I can output for each project:

- if incremental, say something like "Project Foo was previously compiled. Skipping compilation."
- if not incremental, say something like "Project Foo will be compiled. More details on verbose log."

davidfowl commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -370,17 +373,26 @@ private static void AddDependencies(ProjectDependenciesFacade dependencies, Comp
+            compilerIO.Outputs.Add(Path.ChangeExtension(compilationOutput, "pdb"));
+
+            // output: documentation file
+            if (compilerOptions.GenerateXmlDocumentation.HasValue && compilerOptions.GenerateXmlDocumentation.Value)
```
You can do `if (compilerOptions.GenerateXmlDocumentation == true)`

troydai commented on src/Microsoft.DotNet.Tools.Builder/CompileContext.cs:
```diff
@@ -338,23 +338,26 @@ public CompilerIO(List<string> inputs, List<string> outputs)
             compilerIO.Inputs.Add(project.ProjectFile.ProjectFilePath);
 
+            // input: lock file; find when dependencies change
+            compilerIO.Inputs.Add(project.LockFile.LockFilePath);
```
The `LockFile` could be `null`

davidfowl replied:
Add a test for this case

cdmihai replied:
Shouldn't a null lock file be a grave restore error, that ought to make everything shutdown before even getting to build?

davidfowl replied:
It can't be a restore error (since the user may not have run restore yet and thus has no lock file), it *is* a compile error however. Compile handles this but I'm not sure build does. Adding it test can't hurt and it'll be a common scenario.

https://github.com/dotnet/cli/blob/master/src/Microsoft.DotNet.ProjectModel/ProjectContextBuilder.cs#L205

troydai replied:
I think guarding the null here and returning a user friendly message is better than an fatal Null exception.

cdmihai replied:
As David showed, handling a null LockFile already happens in the ProjectContextBuilder with a user friendly message. So I guess the Build verb does not need to handle it again.

I'll add an end to end test that asserts what happens when build is called without a lock file.

troydai commented on src/Microsoft.DotNet.ProjectModel/ProjectContext.cs:
```diff
@@ -22,6 +23,8 @@ public class ProjectContext
 
         public Project ProjectFile => RootProject.Project;
 
+        public LockFile LockFile { get; }
```
We don't have to have this property. The lock file can be found under the project directory. `Path.Combine(ProjectDirectory, LockFile.FileName)`.

davidfowl replied:
This is kinda nice though

cdmihai replied:
@troydai @davidfowl 
Which option do you prefer?

I am in favor of getting the resolved file from the ProjectContext and not duplicating the knowledge on how to find it.

davidfowl replied:
Expose the lock file on ProjectContext

piotrpMSFT commented:
Is this ready to merge?

cdmihai commented:
@piotrpMSFT 

I addressed the feedback; However, some tests are failing on my machine in a most weird manner. :(
Working on debugging the issue. One problem is caused by tests sharing the same intermediary output directory and race on dotnet-compile-csc.rsp

piotrpMSFT commented:
@cdmihai we can't merge until CI is green. If you can get through without addressing these bugs then let's open an issue. if not, we'll need to fix before merging.

piotrpMSFT commented:
@cdmihai if this is intended to go into 1.0, please retarget the PR to rel/1.0.0.

cdmihai commented:
@piotrpMSFT 

I rebased on top of rel/1.0.0 and opened #930.
This PR can either go in master or be closed.

piotrpMSFT commented:
#930 was merged into rel and rel was then merged into master. Closing this.
